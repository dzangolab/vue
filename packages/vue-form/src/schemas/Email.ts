import validator from "validator";
import { z } from "zod";

import type { EmailErrorMessages } from "../types";
import type { IsEmailOptions } from "../types/validator";

const schema = (
  errorMessages: EmailErrorMessages,
  options: IsEmailOptions | undefined
) => {
  return z
    .string({
      required_error: errorMessages.required,
    })
    .superRefine((value, context) => {
      if (value.length === 0) {
        context.addIssue({
          code: "invalid_string",
          message: errorMessages.required,
          validation: "email",
        });
      }

      if (!validator.isEmail(value, options || {})) {
        context.addIssue({
          code: "invalid_string",
          message: errorMessages.invalid,
          validation: "email",
        });
      }
    });
};

export default schema;
