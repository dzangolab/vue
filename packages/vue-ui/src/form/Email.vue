<template>
  <div class="field email">
    <label v-if="label" for="email">
      {{ label }}
    </label>
    <Field
      v-slot="{ field, meta }"
      v-bind="modelValue"
      :name="name"
      :rules="fieldSchema"
    >
      <input
        v-bind="field"
        :class="{
          invalid: meta.touched && !meta.valid,
          valid: meta.dirty && meta.valid,
        }"
        tabindex="0"
        type="email"
      />
      <ErrorMessage :name="name" />
    </Field>
  </div>
</template>

<script lang="ts">
export default {
  name: "Email",
};
</script>

<script setup lang="ts">
import { toFieldValidator } from "@vee-validate/zod";
import validator from "validator";
import { ErrorMessage, Field } from "vee-validate";
import { z } from "zod";

import type { EmailErrorMessages } from "../types/form";
import type { IsEmailOptions } from "../types/validator";
import type { PropType } from "vue";

const props = defineProps({
  errorMessages: {
    default: () => {
      return {
        invalid: "Please provide a valid email address",
        required: "Email address is required",
      };
    },
    required: false,
    type: Object as PropType<EmailErrorMessages>,
  },
  label: {
    default: "",
    required: false,
    type: String as PropType<string>,
  },
  modelValue: {
    default: "",
    type: String as PropType<string | null | undefined>,
  },
  name: {
    default: "email",
    required: false,
    type: String as PropType<string>,
  },
  options: {
    default: () => {
      return {};
    },
    required: false,
    type: Object as PropType<IsEmailOptions>,
  },
});

defineEmits(["update:modelValue"]);

const fieldSchema = toFieldValidator(
  z
    .string({
      required_error: props.errorMessages.required,
    })
    .refine((value) => validator.isEmail(value, props.options), {
      message: props.errorMessages.invalid,
    })
);
</script>
