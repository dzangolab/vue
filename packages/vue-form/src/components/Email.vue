<template>
  <div class="field email">
    <label v-if="label" for="email">
      {{ label }}
    </label>
    <Field
      v-slot="{ field, meta }"
      v-bind="modelValue"
      :name="name"
      :rules="fieldSchema"
    >
      <input
        v-bind="field"
        :class="{
          invalid: meta.touched && !meta.valid,
          valid: meta.dirty && meta.valid,
        }"
        tabindex="0"
        type="email"
      />
      <ErrorMessage :name="name" />
    </Field>
  </div>
</template>

<script lang="ts">
export default {
  name: "Email",
};
</script>

<script setup lang="ts">
import { useI18n } from "@dzangolab/vue3-i18n";
import { toFieldValidator } from "@vee-validate/zod";
import { ErrorMessage, Field } from "vee-validate";

import messages from "../locales/message.json";
import { emailSchema } from "../schemas";

import type { EmailErrorMessages, IsEmailOptions } from "../types";
import type { PropType } from "vue";

const { t } = useI18n({ messages });

const props = defineProps({
  errorMessages: {
    default: () => {
      return {
        invalid: "",
        required: "",
      };
    },
    required: false,
    type: Object as PropType<EmailErrorMessages>,
  },
  label: {
    default: "",
    required: false,
    type: String as PropType<string>,
  },
  modelValue: {
    default: "",
    type: String as PropType<string | null | undefined>,
  },
  name: {
    default: "email",
    required: false,
    type: String as PropType<string>,
  },
  options: {
    default: () => {
      return {};
    },
    required: false,
    type: Object as PropType<IsEmailOptions>,
  },
});

defineEmits(["update:modelValue"]);

const translatedErrorMessages = {
  invalid: props.errorMessages.invalid
    ? props.errorMessages.invalid
    : t("validation.errors.email.invalid"),
  required: props.errorMessages.required
    ? props.errorMessages.required
    : t("validation.errors.email.required"),
};

const fieldSchema = toFieldValidator(
  emailSchema(translatedErrorMessages, props.options)
);
</script>
