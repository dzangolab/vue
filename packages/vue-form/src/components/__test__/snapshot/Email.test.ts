import { mount, RouterLinkStub } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import Email from "../../Email.vue";

describe("Email", () => {
  it("matches snapshot", () => {
    const wrapper = mount(Email, {
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
      props: {
        label: "Email",
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
