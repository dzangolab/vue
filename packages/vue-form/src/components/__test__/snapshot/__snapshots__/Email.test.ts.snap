// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Email > matches snapshot 1`] = `
<div
  class="field email"
>
  <label
    for="email"
  >
    Email
  </label>
  
  <input
    class=""
    name="email"
    placeholder=""
    tabindex="0"
    type="email"
  />
  <!---->
  
</div>
`;
